File: core/ZhiShaoYouKgeZhongFuZiFuDeZuiCO6ww/ZhiShaoYouKgeZhongFuZiFuDeZuiCO6ww.py


def longestSubstring(s: str, k: int) -> int:
    def dfs(left, right):
        cnt = [0] * 26
        for i in range(left, right + 1):
            cnt[ord(s[i]) - ord("a")] += 1

        split = 0
        for i in range(26):
            if 0 < cnt[i] < k:
                split = chr(i + 97)
                break
        if split == 0:
            return right - left + 1

        i, ret = left, 0
        while i <= right:
            while i <= right and s[i] == split:
                i += 1
            if i > right:
                break

            start = i
            while i <= right and s[i] != split:
                i += 1
            length = dfs(start, i - 1)
            ret = max(ret, length)

        return ret

    return dfs(0, len(s) - 1)

if __name__ == '__main__':
    print(longestSubstring(s = "ababbc", k = 2))

File: core/ZhiShaoYouKgeZhongFuZiFuDeZuiCO6ww/ZhiShaoYouKgeZhongFuZiFuDeZuiCO6ww.java
package com.dbc;

public class ZhiShaoYouKgeZhongFuZiFuDeZuiCO6ww {
    private int dfs(String s, int left, int right, int k) {
        int[] cnt = new int[26];
        for (int i = left; i <= right; i++) {
            cnt[s.charAt(i) - 'a']++;
        }

        char split = 0;
        for (int i = 0; i < 26; i++) {
            if (cnt[i] > 0 && cnt[i] < k) {
                split = (char) (i + 'a');
                break;
            }
        }

        if (split == 0) {
            return right - left + 1;
        }

        int i = left, ret = 0;
        while (i <= right) {
            while (i <= right && s.charAt(i) == split) {
                i++;
            }
            if (i > right) {
                break;
            }

            int start = i;
            while (i <= right && s.charAt(i) != split) {
                i++;
            }
            int length = dfs(s, start, i - 1, k);
            ret = Math.max(ret, length);
        }
        return ret;
    }

    public int longestSubstring(String s, int k) {
        return dfs(s, 0, s.length() - 1, k);
    }
}
